### PR TITLE
Add Http json body parser middleware

### PR DESCRIPTION
**Pull request by xavi-belloso**
Hi! 👋
New middleware comes up!

### CODE REVIEW

File: src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Text;
+using Amazon.Lambda.APIGatewayEvents;
+using Newtonsoft.Json;
```
Comment (vgaltes) [inline src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs:4]: I think we´re using system.text.json in other parts of the code. Can we use it here too?

File: src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            }
+           
+            T source;
+            try
+            {
+                source = JsonConvert.DeserializeObject<T>(lambdaEvent.Body);
+            }
+            catch (JsonReaderException)
+            {
+                throw new Exception("Content type defined as JSON but an invalid JSON was provided");
```
Comment (vgaltes) [inline src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs:31]: Why are we throwing a generic exception. Wouldn´t be better to just let the original exception bubble up?

**Review by vgaltes — CHANGES_REQUESTED** [review]
We need to add the same capability for HttpV2 endpoints
We need to add a sample for both cases (rest and httpv2)
We need to update the documentation

File: src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs (hunk 1 above)
Comment (xavi-belloso) [reply src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs]: Yes, already change it

File: src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs (hunk 2 above)
Comment (xavi-belloso) [reply src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/HttpJsonBodyParserMiddleware.cs]: Correct, remove the catch and let original exception bubble up.

File: Voxel.MiddyNet.sln [hunk 3]
```diff
@@ -79,7 +79,15 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Voxel.MiddyNet.ApiGatewayPr
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Voxel.MiddyNet.HttpJsonBodyParserMiddleware.Tests", "test\Voxel.MiddyNet.HttpJsonBodyParserMiddleware.Tests\Voxel.MiddyNet.HttpJsonBodyParserMiddleware.Tests.csproj", "{AD9A0773-EF96-4F4A-890C-337277EDCC2F}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Voxel.MiddyNet.HttpJsonBodyParserMiddleware", "src\Voxel.MiddyNet.HttpJsonBodyParserMiddleware\Voxel.MiddyNet.HttpJsonBodyParserMiddleware.csproj", "{111CB170-E682-498D-9E84-42FFFB66CE36}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Voxel.MiddyNet.HttpJsonBodyParserMiddleware", "src\Voxel.MiddyNet.HttpJsonBodyParserMiddleware\Voxel.MiddyNet.HttpJsonBodyParserMiddleware.csproj", "{111CB170-E682-498D-9E84-42FFFB66CE36}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "HttpJsonBodyParser", "HttpJsonBodyParser", "{2B9304CC-B4DD-4DD5-9428-F7D8A0BC39BC}"
```
Comment (vgaltes) [inline Voxel.MiddyNet.sln:84]: Which project is this one?

File: Voxel.MiddyNet.sln [hunk 4]
```diff
@@ -79,7 +79,15 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Voxel.MiddyNet.ApiGatewayPr
...
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Voxel.MiddyNet.HttpJsonBodyParserMiddleware", "src\Voxel.MiddyNet.HttpJsonBodyParserMiddleware\Voxel.MiddyNet.HttpJsonBodyParserMiddleware.csproj", "{111CB170-E682-498D-9E84-42FFFB66CE36}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Voxel.MiddyNet.HttpJsonBodyParserMiddleware", "src\Voxel.MiddyNet.HttpJsonBodyParserMiddleware\Voxel.MiddyNet.HttpJsonBodyParserMiddleware.csproj", "{111CB170-E682-498D-9E84-42FFFB66CE36}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "HttpJsonBodyParser", "HttpJsonBodyParser", "{2B9304CC-B4DD-4DD5-9428-F7D8A0BC39BC}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Voxel.MiddyNet.HttpJsonBodyParserSample", "samples\HttpJsonBodyParser\Voxel.MiddyNet.HttpJsonBodyParserSample\Voxel.MiddyNet.HttpJsonBodyParserSample.csproj", "{3E45575F-C260-41FD-B5C9-3D675E5BD0C9}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "HttpV2JsonBodyParser", "HttpV2JsonBodyParser", "{308136A6-9610-482A-98E9-D007B6A75437}"
```
Comment (vgaltes) [inline Voxel.MiddyNet.sln:88]: Which project is this one?

**Review by vgaltes — CHANGES_REQUESTED** [review]

File: Voxel.MiddyNet.sln (hunk 3 above)
Comment (xavi-belloso) [reply Voxel.MiddyNet.sln:84]: Is the main Middleware HttpJsonBodyParser project and the other is the tests one

File: Voxel.MiddyNet.sln (hunk 4 above)
Comment (xavi-belloso) [reply Voxel.MiddyNet.sln:88]: The sample project for HttpJsonBodyParserMiddleware

File: docs/source/nuget/packages.rst [hunk 5]
```diff
@@ -264,6 +264,69 @@ And like this for Http API::
         }
     }
 
+Voxel.MiddyNet.HttpJsonBodyParser
+---------------------------------
+This package contains a middleware that parse JSON object mapping to a explicit type. There are two versions avalaible:
```
Comment (vgaltes) [inline docs/source/nuget/packages.rst:269]: This package contains a middleware that parses a JSON to an object of an explicit type.

File: docs/source/nuget/packages.rst [hunk 6]
```diff
@@ -264,6 +264,69 @@ And like this for Http API::
         }
     }
 
+Voxel.MiddyNet.HttpJsonBodyParser
+---------------------------------
+This package contains a middleware that parse JSON object mapping to a explicit type. There are two versions avalaible: 
+
+* One for REST Api (APIGatewayProxyRequest and APIGatewayProxyResponse).
```
Comment (vgaltes) [inline docs/source/nuget/packages.rst:271]: One for REST APIs

File: docs/source/nuget/packages.rst [hunk 7]
```diff
@@ -264,6 +264,69 @@ And like this for Http API::
...
+* And another for Http Api (APIGatewayHttpApiV2ProxyRequest and APIGatewayHttpApiV2ProxyResponse).
```
Comment (vgaltes) [inline docs/source/nuget/packages.rst:272]: One for HTTP APIs

File: docs/source/nuget/packages.rst [hunk 8]
```diff
@@ -264,6 +264,69 @@ And like this for Http API::
...
+
+Configuration
+^^^^^^^^^^^^^
+When you use the middleware, must put the specific type to convert JSON, then the middleware adds a object under context.AdditionalContext with the key "body". To access to the typed object you must do a casting of context["body"] with the custom type.
```
Comment (vgaltes) [inline docs/source/nuget/packages.rst:276]: You need to specify the type you want to convert the JSON object into. The middleware will put the object in the `AdditionalContext["body"]` property. To access the object, you will need to perform an explicit cast of that property.

File: docs/source/nuget/packages.rst [hunk 9]
```diff
@@ -264,6 +264,69 @@ And like this for Http API::
...
+    public class MySample : MiddyNet<APIGatewayProxyRequest, APIGatewayProxyResponse>
+    {
+        public MySample()
+        {
+            Use(new HttpJsonBodyParserMiddleware<Foo>());
+        }
+
+        protected override async Task<APIGatewayProxyResponse> Handle(APIGatewayProxyRequest apiEvent, MiddyNetContext context)
+        {
+            var foo = ((Foo)context.AdditionalContext["Body"]);
```
Comment (vgaltes) [inline docs/source/nuget/packages.rst:291]: In the documentation you say that the property is called "body" (lowercase b)

**Review by vgaltes — CHANGES_REQUESTED** [review]

File: docs/source/nuget/packages.rst (hunk 5 above)
Comment (xavi-belloso) [reply docs/source/nuget/packages.rst]: Thanks!!! 🙏

File: docs/source/nuget/packages.rst (hunk 6 above)
Comment (xavi-belloso) [reply docs/source/nuget/packages.rst]: Thanks!!! 🙏

File: docs/source/nuget/packages.rst (hunk 7 above)
Comment (xavi-belloso) [reply docs/source/nuget/packages.rst]: Done!

File: docs/source/nuget/packages.rst (hunk 8 above)
Comment (xavi-belloso) [reply docs/source/nuget/packages.rst]: Thanks!!! 🙏

File: src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/Constants.cs [hunk 10]
```diff
@@ -0,0 +1,7 @@
+namespace Voxel.MiddyNet.HttpJsonBodyParserMiddleware
+{
+    public class Constants
+    {
+        public const string BodyContextKey = "Body";
```
Comment (vgaltes) [inline src/Voxel.MiddyNet.HttpJsonBodyParserMiddleware/Constants.cs:5]: I would put this directly in the main middleware class, as we´ve done with the TraceContext.

**Review by vgaltes — CHANGES_REQUESTED** [review]

**Review by vgaltes — APPROVED** [review]
Good job!

**Pull request merged**